exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/CreateFunctionHandler.java

```diff
@@ -0,0 +1,195 @@
+        String error = register(remoteRegistry, jar, tmpBinary, tmpSource, remoteRegistry.getRetryTimes());
+
+        if (error != null) {
+          return DirectPlan.createDirectPlan(context, false, error);
+        }
+
+        // remove jars from staging area
+        try {
+          fs.delete(remoteBinary, false);
+          fs.delete(remoteSource, false);
+        } catch (IOException ex) {
+          logger.warn("Failed to delete binary {} and source {} from staging area", jarName, sourceName);
+        }
+
+        return DirectPlan.createDirectPlan(context, true,
+            String.format("The following UDFs in jar %s have been registered:\n%s", jarName, functions));
+
+      } finally {
+        if (inProgress) {
+          remoteRegistry.removeFromJars(jarName);
+        }
+        FileUtils.deleteQuietly(localTmpDir);
+        if (fs.exists(tmpDir)) {
+          fs.delete(tmpDir, true);
+        }
+      }
+    }
+    throw UserException.validationError()
+        .message("Dynamic UDFs support is disabled.")
+        .build(logger);
+  }
+
+  private String validate(Registry registry, Jar jar) {
+    for (Jar remoteJar : registry.getJarList()) {
+      if (remoteJar.getName().equals(jar.getName())) {
+        return String.format("Jar with %s name has been already registered", jar.getName());
+      }
+      for (Func remoteFunction : remoteJar.getFunctionList()) {
+        for (Func func : jar.getFunctionList()) {
+          if (remoteFunction.getName().equals(func.getName()) && remoteFunction.getMajorTypeList().equals(func.getMajorTypeList())) {
+            return String.format("Found duplicated function in %s - %s", remoteJar.getName(), remoteFunction);
+          }
+        }
+      }
+    }
+    return null;
+  }
+
+  private String register(RemoteFunctionRegistry remoteFunctionRegistry, Jar jar, Path tmpBinary, Path tmpSource, int retryTimes) throws IOException {
+    DataChangeVersion version = new DataChangeVersion();
+    Registry registry = remoteFunctionRegistry.getRegistry(version);
+    // validate against remote registry
+    String error = validate(registry, jar);
+    if (error == null) {
+      // copy jars to registry area
+      FileSystem fs = remoteFunctionRegistry.getFs();
+      Path binary = new Path(remoteFunctionRegistry.getRegistryArea(), tmpBinary.getName());
+      Path source = new Path(remoteFunctionRegistry.getRegistryArea(), tmpSource.getName());
+      FileUtil.copy(fs, tmpBinary, fs, binary, false, true, fs.getConf());
+      FileUtil.copy(fs, tmpSource, fs, source, false, true, fs.getConf());
+
+      // add jar info into remote registry if all validation has passed
+      List<Jar> remoteJars = Lists.newArrayList(registry.getJarList());
+      remoteJars.add(jar);
+      Registry updatedRegistry = Registry.newBuilder().addAllJar(remoteJars).build();
+      try {
+        remoteFunctionRegistry.updateRegistry(updatedRegistry, version);
+      } catch (VersionMismatchException ex) {
```

Comment:
Is this how we detect collisions? If so, let's describe the process in a class comment.